torch/fx/passes/utils/matcher_utils.py

```diff
@@ -0,0 +1,230 @@
+
+    # nodes in target graph that are matched placeholder in pattern
+    placeholder_nodes: List[Node] = field(default_factory=list)
+
+    # sink nodes in matched subgraph returned by output
+    returning_nodes: List[Node] = field(default_factory=list)
+
+    def __copy__(self):
+        return _InternalMatch(anchor=self.anchor, nodes_map=self.nodes_map.copy(),
+                              placeholder_nodes=self.placeholder_nodes.copy(),
+                              returning_nodes=self.returning_nodes.copy())
+
+@compatibility(is_backward_compatible=False)
+class SubgraphMatcher:
+    def __init__(self, pattern: Graph,
+                 match_output: bool = False,
+                 match_placeholder: bool = False,
+                 fully_contained: bool = True,
+                 remove_overlapping_matches: bool = True) -> None:
+        """
+        Args:
+            pattern: the targeted matching pattern, represented in fx.Graph
+            match_output: If True, output node in the pattern graph will be treated as a part of the targeted pattern.
+                If False, output node is ignored during match.
+            match_placeholder: If True, placeholder node in the pattern graph will be treated as a part of
+                the targeted pattern. If False, placeholder nodes will be used a wildcard.
+            fully_contained: If True, nodes (except placeholder and output_link_nodes ) can only be consumed
+                by nodes within th pattern.
+            remove_overlapping_matches: If True, in the case of overlapping matches, only the first match
+                will be returned.
+        """
+
+        self.pattern = pattern
+        self.match_output = match_output
+        self.match_placeholder = match_placeholder
+        self.fully_contained = fully_contained
+        self.remove_overlapping_matches = remove_overlapping_matches
+
+        if len(pattern.nodes) == 0:
+            raise ValueError("SubgraphMatcher cannot be initialized with an empty pattern")
+
+        for node in pattern.nodes:
+            if node.op != "output":
+                assert len(node.users) > 0, \
+                       "SubgraphMatcher cannot be initialized with an pattern with dead code"
+
+        # TODO: assert pattern is a connected graph
+
+        self.pattern_placeholder_nodes = [n for n in pattern.nodes if n.op == "placeholder"]
+        output_node = next(iter(reversed(pattern.nodes)))
+        # nodes returned by outputs
+        self.pattern_returning_nodes: List[Node] = output_node.all_input_nodes
+
+        self.pattern_anchors: List[Node] = []
+        if match_output:
+            self.pattern_anchors = [output_node]
+        else:
+            # If a node has output_node as the ONLY user, then this node is a graph sink,
+            # and should be matched against as an anchor
+            self.pattern_anchors = [n for n in output_node.all_input_nodes if len(n.users) == 1]
```

Comment:
Pattern anchor initialization here doesn't look quite right. If the pattern is `fully_contained`, this looks right. But if it is not, all of the internal nodes need to be anchors too. That being said, if you dump `fully_contained == False` case as I suggested above you won't have to deal with this case, it will automatically do the right thing.

Reply:
I removed the fully_contained == False  case.